docs/design/proposals/encryption-with-vault-sa.md

```diff
@@ -0,0 +1,84 @@
+one of the implementations can be added.
+
+Different KMS implementations and their configurable options can be found at
+[`csi-kms-connection-details.yaml`](https://github.com/ceph/ceph-csi/blob/devel/examples/kms/vault/csi-kms-connection-details.yaml).
+
+### VaultTokensKMS
+
+- global configuration in the storage deployment namespace
+- optional configuration per Tenant, allows for overriding Vault options
+- fetches a Secret from the Tenants (volume owner) namespace
+
+An example of the per Tenant configuration options are in
+[`tenant-config.yaml`](https://github.com/ceph/ceph-csi/blob/devel/examples/kms/vault/tenant-config.yaml)
+and
+[`tenant-token.yaml`](https://github.com/ceph/ceph-csi/blob/devel/examples/kms/vault/tenant-token.yaml).
+
+Implementation is in [`vault_tokens.go`](https://github.com/ceph/ceph-csi/blob/devel/internal/util/vault_tokens.go).
+
+### Vault
+
+- uses Kubernetes Auth with single service account (aka storage admin account)
+
+Implementation is in [`vault.go`](https://github.com/ceph/ceph-csi/blob/devel/internal/util/vault.go).
+
+## Extension or New KMS implementation
+
+Normally ServiceAccounts are provided by Kubernetes in the containers
+filesystem. This only allowas a single ServiceAccount and is static for the
+lifetime of the Pod. Ceph-CSI runs in the namespace of the storage
+administrator, and has access to the single ServiceAccount linked in the
+deployments of the Pods.
+
+For Ceph-CSI to dynamically use ServiceAccounts from tenants, the following
+steps need to be taken:
+
+- use the owner of the volume to identify the tenant
+- identify the right ServiceAccount in the tenants namespace
+- read and parse the ServiceAccount contents
+- provide the ServiceAccount contents as a (temporary) directory structure
+- setup the Vault connection to use the contents from the ServiceAccount to
+  replace the default (`AuthKubernetesTokenPath:
+  /var/run/secrets/kubernetes.io/serviceaccount/token`)
+
+Currently the Ceph-CSI components may read Secrets and ConfigMaps from the
+Tenants namespace. These permissions need to be extended to allow Ceph-CSI to
+read the contents of the ServiceAccount(s) in the Tenants namespace.
+
+## High-Level Design
+
+### Global Configuration
+
+1. a StorageClass links to a KMS configuration by providing the `kmsID` parameter
+1. a ConfigMap in the namespace of the Ceph-CSI deployment contains the KMS
+   configuration for the `kmsID`
+   ([`csi-kms-connection-details.yaml`](https://github.com/ceph/ceph-csi/blob/devel/examples/kms/vault/csi-kms-connection-details.yaml))
+
+When Ceph-CSI receives a `VolumeCreate` request, the parameters from the
+StorageClass and details about the requested Volume are available. The Ceph-CSI
+provisioner checks the `kmsID` parameter and fetches the matching KMS
+configuration from the ConfigMap.
+
+### Tenant Configuration
+
+1. needs an identifiable ServiceAccount with permissions to connect to Vault
```

Comment:
just trying to figure out what is meant by `identifiable` here ?

Reply:
I've replaced it with "ServiceAccount with a known name" now, hopefully that clarifies it